crates/adapters/src/transport/durable/kafka/input.rs

```diff
@@ -0,0 +1,771 @@
+            } else if req_step >= steps.end {
+                break;
+            }
+
+            // Read the data offsets for the step.
+            let step_message = read_partition_queue(&consumer, &step_partition)?;
+            let payload = step_message.payload().unwrap_or(&[]);
+            let step_offsets = serde_json::from_slice::<IndexEntry>(payload)?;
+            assert_eq!(step_offsets.step, step);
+            assert!(step_offsets.data_offsets.start >= data_offset);
+
+            // Read the data for the step.
+            receiver.soi(step, true);
+            while data_offset < step_offsets.data_offsets.end {
+                let data_message = read_partition_queue(&consumer, &data_partition)?;
+                if data_message.offset() < step_offsets.data_offsets.start {
+                    continue;
+                }
+                if let Some(payload) = data_message.payload() {
+                    let len = step_offsets.data_offsets.end - step_offsets.data_offsets.start;
+                    if len == 1 {
+                        let _ = receiver.input_chunk(payload);
+                    } else {
+                        let _ = receiver.input_fragment(payload);
+                    }
+                }
+                data_offset = data_message.offset() + 1;
+            }
+            receiver.eoi();
+            receiver.settled();
+
+            step += 1;
+        }
+
+        debug!("Creating Kafka producer");
+        let (producer_sender, producer_receiver) = channel();
+        let context = KafkaInputContext::new(producer_sender);
+        let step_producer =
+            ThreadedProducer::from_config_and_context(&inner.producer_config, context)?;
+
+        // Fetch the watermarks for the topic we're going to produce to, and
+        // then throw away the information because we don't care.  librdkafka
+        // needs this information, and it will fetch it for itself eventually,
+        // but it likes to wait 1 full second before asking for it and fetching
+        // it immediately avoids that delay.
+        step_producer
+            .client()
+            .fetch_watermarks(index_topic, inner.partition, None)?;
+
+        loop {
+            poll_consumer(&consumer)?;
+
+            let req_step = match *state.lock().unwrap() {
+                State::Start => unreachable!(),
+                State::Read(_first, step) => step,
+                State::Exit => return Ok(()),
+            };
+            if req_step < step {
+                parker.park_timeout(POLL_TIMEOUT);
+                continue;
+            };
+            let force = req_step > step;
+
+            // Now create `step`.
+            let start_offset = data_offset;
+            if force {
+                receiver.soi(step, false);
+            } else {
+                let Some(result) = data_partition.poll(POLL_TIMEOUT) else {
+                    continue;
+                };
+                let data_message = result?;
```

Comment:
Can this be a transient error?
If bailing out is the best option here, how will this error propagate to the main thread?